PubSub/tests/Unit/BatchPublisherTest.php

```diff
@@ -148,6 +148,72 @@ public function testPublishDeferred()
         $res = $publisher->publishDeferred($messages);
         $this->assertEquals(array_fill(0, count($messages), 1), $res);
     }
+
+    /**
+     * @dataProvider compressionDataProvider
+     */
+    public function testPublisherCompression(
+        $enableCompression,
+        $compressionBytesThreshold,
+        $shouldCompress
+    ) {
+        $client = TestHelpers::stub(PubSubClient::class, [
+            ['suppressKeyFileNotice' => true, 'projectId' => 'example-project']
+        ], [
+            'encode', 'connection'
+        ]);
+        $client->___setProperty('encode', false);
+
+        $publisher = TestHelpers::stub(BatchPublisher::class, [
+            self::TOPIC_NAME, [
+                'identifier' => uniqid(),
+                'enableCompression' => $enableCompression,
+                'compressionBytesThreshold' => $compressionBytesThreshold
+            ]
+        ], ['client', 'topics']);
+
+        $connection = $this->prophesize(ConnectionInterface::class);
+
+        $messages = [[
+            'data' => 'foo'
+        ]];
+
+        $compressionHeader = [];
+        if ($shouldCompress) {
+            $compressionHeader = [
+                'grpc-internal-encoding-request' => [['gzip']]
+            ];
+        }
+
+        $connection->publishMessage(Argument::that(
+            function ($args) use ($compressionHeader) {
+                $result = is_array($args) && array_key_exists('messages', $args);
+                foreach ($compressionHeader as $key => $value) {
+                    $result = $result &&
+                        isset($args['headers']) &&
+                        array_key_exists($key, $args['headers']) &&
+                        $args['headers'][$key] == $value;
+                }
+                return $result;
+            }
+        ))->shouldBeCalled(1)->willReturn([]);
+
+        $client->___setProperty('connection', $connection->reveal());
+        $publisher->___setProperty('client', $client);
+        $publisher->publishDeferred($messages);
+    }
+
+    public function compressionDataProvider()
+    {
+        // Each data is of the form
+        // [$enableCompression, $compressionBytesThreshold, $shouldCompress]
+        return [
+            [false, 0, false],
+            [false, 10000, false],
+            [true, 0, true],
```

Comment:
Do you want to add a case when `$shouldCompress` is false but `$compressionBytesThreshold` is provided?
Ex: `true, <greater than zero>, false`

Should it fail in that case?

Reply:
I'm sorry I didn't get this case. Let me illustrate the test case so that we are on the same page.

`$enableCompression` and `$compressionBytesThreshold` are properties which would eventually be set while creating the `batchPublisher`. The **`$shouldCompression`** variable is for the test to know whether compression should actually happen by the library for that particular compression options and message.

Reply:
Oh yes, makes sense.